### PR TITLE
rustdoc: Rename "object safe" to "dyn compatible"

### PR DESCRIPTION
**Pull request by fmease**
Supersedes #126554:

1. In line with [T-lang's latest resolution](https://github.com/rust-lang/lang-team/issues/286#issuecomment-2338905118).
2. More comprehensive: Not only updates user-facing text but also source code.

Part of #130852.

Doesn't update rustdoc-JSON (~~will be~~ filed separately. #131595).

r? @notriddle (rust-lang/lang-team#286) @GuillaumeGomez (for visibility)

### CODE REVIEW

File: src/librustdoc/html/render/print_item.rs [hunk 1]
```diff
@@ -934,16 +934,17 @@ fn item_trait(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, t: &clean:
...
-            "object-safety",
+            "Dyn Compatibility",
+            "dyn-compatibility",
             None,
             &format!(
-                "<div class=\"object-safety-info\">This trait is <b>not</b> \
+                "<div class=\"dyn-compatibility-info\">This trait is <b>not</b> \
                 <a href=\"{base}/reference/items/traits.html#object-safety\">\
-                object safe</a>.</div>",
+                dyn compatible</a> (formerly known as object safe).</div>",
```
Comment (fmease) [inline src/librustdoc/html/render/print_item.rs:947]: Let me know if you know of a 'better way' to present the "formerly known" part. For context, in the compiler & library docs, I went for MarkDown footnotes. I wonder if we should quote or italicize "object safe".

File: src/librustdoc/html/render/print_item.rs [hunk 2]
```diff
@@ -934,16 +934,17 @@ fn item_trait(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, t: &clean:
     let cache = &cloned_shared.cache;
     let mut extern_crates = FxIndexSet::default();
 
-    if !t.is_object_safe(cx.tcx()) {
+    if !t.is_dyn_compatible(cx.tcx()) {
+        // FIXME(dyn_compat_renaming): Update URL once the Reference is updated and hits stable.
```
[the remark this answers is not in the document]
Comment (fmease) [reply src/librustdoc/html/render/print_item.rs:938]: Ah good call, thanks! I didn't look carefully enough

**Comment by GuillaumeGomez** [comment]
Apart from notriddle's comments, looks good to me. :)

**Comment by fmease** [comment]
Suggestions applied. Ready.

**Pull request merged**